__version__ = "1.0"

import os
from os.path import expanduser
import warnings


def get_default_root():
    home = expanduser("~")
    dirpath = os.path.join(home, ".mpramnist", "data", "")

    try:
        if not os.path.exists(dirpath):
            os.makedirs(dirpath)
    except:
        warnings.warn("Failed to setup default root.")
        dirpath = None

    return dirpath


DEFAULT_ROOT = get_default_root()

HOMEPAGE = "https://github.com/autosome-imtf/MPRA-MNIST"

INFO = {
    "Agarwal": {
        "python_class": "Agarwal",
        "description": "The Agarwal dataset is based on a lentiviral MPRA system. The total data volume was 122,926 sequences for the HepG2 cell line, 196,664 for K562, and 46,185 for WTC11. Each sequence is 200 nucleotides long. The regression task was to predict a scalar value of regulatory activity for the corresponding cell line.",
        "url_Agarwal_HepG2.tsv": "https://zenodo.org/records/17453603/files/Agarwal_HepG2.tsv?download=1",
        "MD5_Agarwal_HepG2.tsv": "13766f66e19d57c250fe733f4f521bc9",
        "url_Agarwal_K562.tsv": "https://zenodo.org/records/17453603/files/Agarwal_K562.tsv?download=1",
        "MD5_Agarwal_K562.tsv": "27e61525d4344da4d40f60cceadb6069",
        "url_Agarwal_WTC11.tsv": "https://zenodo.org/records/17453603/files/Agarwal_WTC11.tsv?download=1",
        "MD5_Agarwal_WTC11.tsv": "831ac9bfe38d103e67492dc3a3618ff8",
        "target_columns": {
            "expression": "Column for expression",
            "averaged_expression": "Average activity of forward and reverse-complement sequences",
        },
        "scalar_features": {},
        "vector_features": {},
        "task": "Regression",
        "split": {"train": "1, 2, 3, 4, 5, 6, 7, 8", "val": 9, "test": 10},
        "Agarwal_HepG2_split": {"train": 98336, "val": 12298, "test": 12298},
        "Agarwal_K562_split": {"train": 167328, "val": 19670, "test": 19670},
        "Agarwal_WTC11_split": {"train": 36946, "val": 4622, "test": 4622},
    },
    "AgarwalJoint": {
        "python_class": "AgarwalJoint",
        "description": "The AgarwalJoint dataset contains 55,338 sequences, each 200 nucleotides long, comprising enhancers from the HepG2, K562, and WTC11 cell lines. The regression task is to predict three scalar values of regulatory activity for each of these cell lines.",
        "url_AgarwalJoint_joint_data.tsv": "https://zenodo.org/records/17453603/files/AgarwalJoint_joint_data.tsv?download=1",
        "MD5_AgarwalJoint_joint_data.tsv": "1112a31cbd48e46d4656fcb9abc22387",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {"HepG2", "K562", "WTC11"},
        "task": "Regression",
        "split": {"train": "1, 2, 3, 4, 5, 6, 7, 8", "val": 9, "test": 10},
        "AgarwalJoint_split": {"train": 36946, "val": 4622, "test": 4622},
    },
    "Kircher": {
        "python_class": "Kircher",
        "description": "The Kircher dataset is based on MPRA saturation mutagenesis results and includes 44,647 variants of regulatory elements tested in various cell lines. These data are proposed to be used as a benchmark set for evaluating the quality of machine learning models. The regression task involved predicting a scalar value of regulatory activity for the corresponding cell line.",
        "url_Kircher_GRCh38_ALL.tsv": "https://zenodo.org/records/17453603/files/Kircher_GRCh38_ALL.tsv?download=1",
        "MD5_Kircher_GRCh38_ALL.tsv": "eacbb118660a3bd39575d88b1a03f7a7",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {},
        "task": "Regression",
        "split": {},
        "Kircher_split": {
            'HepG2': ['F9', 'LDLR', 'LDLR.2', 'SORT1', 'SORT1-flip', 'SORT1.2'],
            'HeLa': ['FOXE1'],
            'HEL92.1.7': ['GP1BA', 'HBB', 'HBG1', 'BCL11A'],
            'HEK293T': ['HNF4A', 'MSMB', 'TERT-HEK', 'MYCrs6983267'],
            'K562': ['PKLR-24h', 'PKLR-48h'],
            'SF7996': ['TERT-GAa', 'TERT-GBM', 'TERT-GSc'],
            'SK-MEL-28': ['IRF4'],
            'HaCaT': ['IRF6'],
            'LNCaP': ['MYCrs11986220'],
            'Neuro-2a': ['RET', 'UC88'],
            'MIN6': ['TCF7L2', 'ZFAND3'],
            'NIH-3T3': ['ZRSh-13', 'ZRSh-13h2']
        },
        "Kircher_construct_size": {
            "BCL11A": 600,
            "FOXE1": 600,
            "MSMB": 593,
            "ZFAND3": 579,
            "SORT1.2": 600,
            "SORT1-flip": 600,
            "UC88": 590,
            "RET": 600,
            "MYCrs6983267": 600,
            "SORT1": 600,
            "TCF7L2": 600,
            "IRF6": 600,
            "PKLR-48h": 470,
            "PKLR-24h": 470,
            "MYCrs11986220": 464,
            "ZRSh-13h2": 485,
            "ZRSh-13": 485,
            "IRF4": 451,
            "GP1BA": 385,
            "LDLR.2": 318,
            "LDLR": 318,
            "F9": 303,
            "HNF4A": 285,
            "TERT-HEK": 259,
            "TERT-GAa": 259,
            "TERT-GSc": 259,
            "TERT-GBM": 259,
            "HBG1": 274,
            "HBB": 187,
        },
    },
    "Malinois": {
        "python_class": "Malinois",
        "description": "The Malinois dataset includes 798,064 sequences tested in the K562, HepG2, and SK-N-SH cell lines. The original sequence length is approximately 200 nucleotides, and it is recommended to extend them to 600 bp. The task is to predict three normalized regulatory activity values for the respective cell lines.",
        "url_Malinois_Table_S2.tsv": "https://zenodo.org/records/17453603/files/Malinois_Table_S2.tsv?download=1",
        "MD5_Malinois_Table_S2.tsv": "6b17ac4091c28ead54cf962ed557b3b4",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {"K562_log2FC", "HepG2_log2FC", "SKNSH_log2FC"},
        "task": "Regression",
        "split": {
            "train": "1, 2, 3, 4, 5, 6, 8, 9, 10, 11, 12, 14, 15, 16, 17, 18, 20, 22, Y",
            "val": "19, 21, X",
            "test": "7, 13",
        },
        "Malinois_split": {"train": 668946, "val": 58809, "test": 62582},
    },
    "StarrSeq": {
        "python_class": "StarrSeq",
        "description": "The STARR-seq dataset is based on several libraries designed to determine the activity of human regulatory elements. The classification task for most libraries was to predict whether a sequence is an active or inactive regulatory element.",
        "StarrSeq_ATACSeq_description": "ATAC Seq: Active elements were defined by ATAC-seq peaks.",
        "StarrSeq_binary_description": "Binary: Contains promoter-enhancer pairs for GP5d and HepG2 cell lines. For the 'Binary' dataset, the task was to predict active and inactive promoter-enhancer pairs.",
        "StarrSeq_CaptProm_description": "Capture Promoter: Determines promoter activity in GP5d cells.",
        "StarrSeq_genEnh_description": "Genomic Enhancer: Consists of genomic DNA fragments.",
        "StarrSeq_genProm_description": "Genomic Promoter: Determines promoter activity in GP5d cells.",
        "StarrSeq_ranEnh_description": "Random Enhancer: Used to identify enhancer features in GP5d cells.",
        "url_StarrSeq_ATACSeq_all_chr_file.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_ATACSeq_all_chr_file.fasta.gz?download=1",
        "MD5_StarrSeq_ATACSeq_all_chr_file.fasta.gz": "739c90c58c93194a5df818d98d00691e",
        "url_StarrSeq_binary_test_enhancer.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_test_enhancer.fasta.gz?download=1",
        "MD5_StarrSeq_binary_test_enhancer.fasta.gz": "ed01b85d47506b64eb464bb4b09e389b",
        "url_StarrSeq_binary_test_promoter.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_test_promoter.fasta.gz?download=1",
        "MD5_StarrSeq_binary_test_promoter.fasta.gz": "c7561f6bc977693079d38e00f22378ce",
        "url_StarrSeq_binary_train_enhancer.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_train_enhancer.fasta.gz?download=1",
        "MD5_StarrSeq_binary_train_enhancer.fasta.gz": "3e474ac6c6d87a05859ada8e5b7dd1c1",
        "url_StarrSeq_binary_train_enhancer_from_input.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_train_enhancer_from_input.fasta.gz?download=1",
        "MD5_StarrSeq_binary_train_enhancer_from_input.fasta.gz": "8d72efc718a314d63b77b46a6da8ec79",
        "url_StarrSeq_binary_train_enhancer_permutated.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_train_enhancer_permutated.fasta.gz?download=1",
        "MD5_StarrSeq_binary_train_enhancer_permutated.fasta.gz": "c347a66f16211559e049de9c4e80c33d",
        "url_StarrSeq_binary_train_promoter.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_train_promoter.fasta.gz?download=1",
        "MD5_StarrSeq_binary_train_promoter.fasta.gz": "90173f2f042416e2428105b2920b98b8",
        "url_StarrSeq_binary_train_promoter_from_input.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_train_promoter_from_input.fasta.gz?download=1",
        "MD5_StarrSeq_binary_train_promoter_from_input.fasta.gz": "8494ca5292a6d9445052eb99cc856b00",
        "url_StarrSeq_binary_val_enhancer.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_val_enhancer.fasta.gz?download=1",
        "MD5_StarrSeq_binary_val_enhancer.fasta.gz": "ede1ad8bec2f633e99b7e6f9d6866981",
        "url_StarrSeq_binary_val_promoter.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_binary_val_promoter.fasta.gz?download=1",
        "MD5_StarrSeq_binary_val_promoter.fasta.gz": "1b7a6177120a4f875f3374ba82c8e6f0",
        "url_StarrSeq_CaptProm_test.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_CaptProm_test.fasta.gz?download=1",
        "MD5_StarrSeq_CaptProm_test.fasta.gz": "9596d4755fe6ddf0039d81603106218d",
        "url_StarrSeq_CaptProm_train.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_CaptProm_train.fasta.gz?download=1",
        "MD5_StarrSeq_CaptProm_train.fasta.gz": "e0235ba235742d362e8c4b365a036628",
        "url_StarrSeq_CaptProm_val.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_CaptProm_val.fasta.gz?download=1",
        "MD5_StarrSeq_CaptProm_val.fasta.gz": "40aa4578124ceeb09c84c76f51c7d0bd",
        "url_StarrSeq_genEnh_all_chr_file.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_genEnh_all_chr_file.fasta.gz?download=1",
        "MD5_StarrSeq_genEnh_all_chr_file.fasta.gz": "6332a6dd899f24d6612de4963af6c52e",
        "url_StarrSeq_genProm_test.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_genProm_test.fasta.gz?download=1",
        "MD5_StarrSeq_genProm_test.fasta.gz": "b06811e1fff61dfaaee0cac17a453b35",
        "url_StarrSeq_genProm_train.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_genProm_train.fasta.gz?download=1",
        "MD5_StarrSeq_genProm_train.fasta.gz": "b52afbe57a66e4c4f4381a4e8ffb291e",
        "url_StarrSeq_genProm_val.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_genProm_val.fasta.gz?download=1",
        "MD5_StarrSeq_genProm_val.fasta.gz": "5f783a547346be2f0ba41530a42125da",
        "url_StarrSeq_ranEnh_test.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_ranEnh_test.fasta.gz?download=1",
        "MD5_StarrSeq_ranEnh_test.fasta.gz": "0bffb1fa925f676f5b0b20d7f3aa12fe",
        "url_StarrSeq_ranEnh_train.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_ranEnh_train.fasta.gz?download=1",
        "MD5_StarrSeq_ranEnh_train.fasta.gz": "1d90ec94f6d69ae74ad36da4ff046e1e",
        "url_StarrSeq_ranEnh_val.fasta.gz": "https://zenodo.org/records/15195607/files/StarrSeq_ranEnh_val.fasta.gz?download=1",
        "MD5_StarrSeq_ranEnh_val.fasta.gz": "ab00ecd6dbeb1e494a40d64cb9719735",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {},
        "task": "Binary Classification",
        "split": {"train", "val", "test"},
        "StarrSeq_ATACSeq_split": {"train": 2712578, "val": 910186, "test": 912794},
        "StarrSeq_binary_split": {"train": 2991302, "val": 748828, "test": 1252044},
        "StarrSeq_CaptProm_split": {"train": 79732, "val": 13290, "test": 13290},
        "StarrSeq_genEnh_split": {"train": 323756, "val": 89062, "test": 99362},
        "StarrSeq_genProm_split": {"train": 36898, "val": 4098, "test": 7285},
        "StarrSeq_ranEnh_split": {"train": 7942528, "val": 1985634, "test": 3309387},
    },
    "Sure": {
        "python_class": "Sure",
        "description": "The Sure dataset contains data from the analysis of genomes from 4 individuals. Experiments were conducted on the K562 and HepG2 cell lines. The dataset volume is ~400–600 thousand training and ~50–70 thousand test and validation sequences per individual. The classification task is to predict a class label reflecting the expression level for each cell line. The regression task is to predict a continuous value of the expression level for K562 and HepG2.",
        "url_Sure_SuRE42_HG02601.tsv": "https://zenodo.org/records/17453603/files/Sure_SuRE42_HG02601.tsv?download=1",
        "MD5_Sure_SuRE42_HG02601.tsv": "ca3cf305a4a203d19037b1776b123108",
        "url_Sure_SuRE43_GM18983.tsv": "https://zenodo.org/records/17453603/files/Sure_SuRE43_GM18983.tsv?download=1",
        "MD5_Sure_SuRE43_GM18983.tsv": "34c4134feeea5df42296c54be179f56e",
        "url_Sure_SuRE44_HG01241.tsv": "https://zenodo.org/records/17453603/files/Sure_SuRE44_HG01241.tsv?download=1",
        "MD5_Sure_SuRE44_HG01241.tsv": "4769f18bf2ab7d45e4a6ee0c0eeda985",
        "url_Sure_SuRE45_HG03464.tsv": "https://zenodo.org/records/17453603/files/Sure_SuRE45_HG03464.tsv?download=1",
        "MD5_Sure_SuRE45_HG03464.tsv": "e042c18c2a0890df49423d77e87dc26c",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {
            "classification": ["K562_bin", "HepG2_bin"],
            "regression": ["avg_K562_exp", "avg_HepG2_exp"],
        },
        "task": "Regression or Multi-class",
        "split": {"train", "val", "test"},
        "Sure_SuRE42_HG02601_split": {"train": 619051, "val": 73935, "test": 73947},
        "Sure_SuRE43_GM18983_split": {"train": 554829, "val": 72758, "test": 72774},
        "Sure_SuRE44_HG01241_split": {"train": 446214, "val": 62580, "test": 62639},
        "Sure_SuRE45_HG03464_split": {"train": 385977, "val": 50331, "test": 50422},
    },
    "Sharpr": {
        "python_class": "Sharpr",
        "description": "The Sharpr dataset contains activity measurements for approximately 477 thousand synthetic promoter sequences, each 145 bp long, tested in K562 and HepG2 cell lines. About 427 thousand sequences are used for training. The task is to predict 12 scalar values representing gene expression activity measurements.",
        "url_Sharpr_test.tsv": "https://zenodo.org/records/15195607/files/Sharpr_test.tsv?download=1",
        "MD5_Sharpr_test.tsv": "c93d3bd2b5a9e8e8f71699af50a81614",
        "url_Sharpr_train.tsv": "https://zenodo.org/records/15195607/files/Sharpr_train.tsv?download=1",
        "MD5_Sharpr_train.tsv": "b862d888630bda8de8f968fa930e2203",
        "url_Sharpr_val.tsv": "Sharpr_val.tsv",
        "MD5_Sharpr_val.tsv": "c93d3bd2b5a9e8e8f71699af50a81614",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {
            "k562_minp_rep1",
            "k562_minp_rep2",
            "k562_minp_avg",
            "k562_sv40p_rep1",
            "k562_sv40p_rep2",
            "k562_sv40p_avg",
            "hepg2_minp_rep1",
            "hepg2_minp_rep2",
            "hepg2_minp_avg",
            "hepg2_sv40p_rep1",
            "hepg2_sv40p_rep2",
            "hepg2_sv40p_avg",
        },
        "task": "Multilabel",
        "split": {"train", "val", "test"},
        "Sharpr_split": {"train": 457174, "val": 10130, "test": 10130},
    },
    "Fluorescence": {
        "python_class": "Fluorescence",
        "description": "The Fluorescence dataset is based on a lentiviral MPRA system, collected for three immune cell lines: Jurkat, K562, and THP-1. The dataset includes 17,104 sequences, each 250 bp long. The regression task was to predict normalized regulatory activity values for each cell line.",
        "url_Fluorescence_test.tsv": "https://zenodo.org/records/15195607/files/Fluorescence_test.tsv?download=1",
        "MD5_Fluorescence_test.tsv": "23dc89727366d031f7d55506585ed97f",
        "url_Fluorescence_train.tsv": "https://zenodo.org/records/15195607/files/Fluorescence_train.tsv?download=1",
        "MD5_Fluorescence_train.tsv": "703e7831688549aeab4b391988dade84",
        "url_Fluorescence_val.tsv": "https://zenodo.org/records/15195607/files/Fluorescence_val.tsv?download=1",
        "MD5_Fluorescence_val.tsv": "9d7f564bb9cc4f9100edf784fe6d1001",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {"JURKAT", "K562", "THP1"},
        "task": "Regression",
        "split": {"train", "val", "test"},
        "Fluorescence_split": {"train": 12335, "val": 1416, "test": 3353},
    },
    "Evfratov": {
        "python_class": "Evfratov",
        "description": "The Evfratov dataset is based on a study of the influence of 5′-UTR sequences on translation efficiency. The dataset contains 11,692 sequences of 23 nucleotides in length and 11,889 sequences of 33 nucleotides in length. The data enables the task of multiclass classification, where each of the 8 classes corresponds to a range of translation efficiency.",
        "url_Evfratov_23_test.tsv": "https://zenodo.org/records/15195607/files/Evfratov_23_test.tsv?download=1",
        "MD5_Evfratov_23_test.tsv": "a1bfaced916f2cd1fd312eca4aa36787",
        "url_Evfratov_23_train.tsv": "https://zenodo.org/records/15195607/files/Evfratov_23_train.tsv?download=1",
        "MD5_Evfratov_23_train.tsv": "3bccff4b46d3a859f1bcaa3c4204bebd",
        "url_Evfratov_23_val.tsv": "https://zenodo.org/records/15195607/files/Evfratov_23_val.tsv?download=1",
        "MD5_Evfratov_23_val.tsv": "174fe1f42aab6888e1a1e90d70b89bd6",
        "url_Evfratov_33_test.tsv": "https://zenodo.org/records/15195607/files/Evfratov_33_test.tsv?download=1",
        "MD5_Evfratov_33_test.tsv": "b3af0b99fbe1e0a1400935294cff9520",
        "url_Evfratov_33_train.tsv": "https://zenodo.org/records/15195607/files/Evfratov_33_train.tsv?download=1",
        "MD5_Evfratov_33_train.tsv": "622674c09d2e55c6b783e1fc8eedfab2",
        "url_Evfratov_33_val.tsv": "https://zenodo.org/records/15195607/files/Evfratov_33_val.tsv?download=1",
        "MD5_Evfratov_33_val.tsv": "cb23a2c03f0d690f5d0abe233d653de2",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {},
        "task": "Multi-class",
        "split": {"train", "val", "test"},
        "Evfratov_23_split": {"train": 9353, "val": 1169, "test": 1170},
        "Evfratov_33_split": {"train": 9511, "val": 1189, "test": 1189},
    },
    "DeepPromoter": {
        "python_class": "DeepPromoter",
        "description": "The DeepPromoter dataset contains 14,098 E. coli promoters. Each sequence is 50 nucleotides long. The regression task was to predict a scalar value of regulatory activity.",
        "url_DeepPromoter_all_seqs.tsv": "https://zenodo.org/records/15195607/files/DeepPromoter_all_seqs.tsv?download=1",
        "MD5_DeepPromoter_all_seqs.tsv": "242bac7cbe193700b680bb6e839a3f06",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {},
        "task": "Regression",
        "split": {"train", "val", "test"},
        "DeepPromoter_split": {"train": 9000, "val": 1000, "test": 1184},
    },
    "Dream": {
        "python_class": "Dream",
        "description": "The Dream dataset is based on the DREAM challenge and contains data on the activity of promoters in the yeast Saccharomyces cerevisiae. The dataset comprises 6.7 million sequences. The input data consists of DNA sequences of variable length, and the output is a scalar value of regulatory activity. The task is to predict this value.",
        "url_Dream_paired.tsv": "https://zenodo.org/records/15195607/files/Dream_paired.tsv?download=1",
        "MD5_Dream_paired.tsv": "7512ae3f91711af4331210548cb3cb52",
        "url_Dream_single.tsv": "https://zenodo.org/records/15195607/files/Dream_single.tsv?download=1",
        "MD5_Dream_single.tsv": "1bcbc80702d74151c57e2da8a82b2c7a",
        "url_Dream_train.tsv": "https://zenodo.org/records/15195607/files/Dream_train.tsv?download=1",
        "MD5_Dream_train.tsv": "a7f71419d75b048a1f1b84c352b3ab78",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {},
        "task": "Regression",
        "split": {"train", "val", "test"},
        "Dream_split": {
            "train": 6739258,
            "val": {
                "all sequences": 9045,
                "high": 96,
                "low": 99,
                "yeast": 208,  # == native
                "random": 733,
                "challenging": 195,
                "snv": 4670,  # paired types
                "perturbation": 328,
                "tiling": 263,
            },
            "test": {
                "all sequences": 62058,
                "high": 872,
                "low": 898,
                "yeast": 370,  # == native
                "random": 5616,
                "challenging": 1758,
                "snv": 39904,  # paired types
                "perturbation": 2959,
                "tiling": 2361,
            },
        },
    },
    "Vaishnav": {
        "python_class": "Vaishnav",
        "description": "The Vaishnav dataset includes measurements of promoter-dependent expression: 30,722,376 for the standard medium (YPD) and 20,616,659 for the minimal medium (SD-Ura). The input data are DNA sequences of variable length, and the outputs are a scalar value of regulatory activity. The data is divided into training and validation sets in a 9:1 ratio. The independent test set contains measurements for native promoters: 3,928 for YPD and 3,977 for SD-Ura.",
        "url_Vaishnav_complex_drift.tsv": "https://zenodo.org/records/15195607/files/Vaishnav_complex_drift.tsv?download=1",
        "MD5_Vaishnav_complex_drift.tsv": "26eb589f80e6f7922d4fdcc561ff5e52",
        "url_Vaishnav_complex_native.tsv": "https://zenodo.org/records/15195607/files/Vaishnav_complex_native.tsv?download=1",
        "MD5_Vaishnav_complex_native.tsv": "d18cec6b7fa58da1c5f761f6551804d8",
        "url_Vaishnav_complex_paired.tsv": "https://zenodo.org/records/15195607/files/Vaishnav_complex_paired.tsv?download=1",
        "MD5_Vaishnav_complex_paired.tsv": "52f1f353ecdcbee7ad7a6f930ec9150d",
        "url_Vaishnav_complex_train_val.tsv": "https://zenodo.org/records/15195607/files/Vaishnav_complex_train_val.tsv?download=1",
        "MD5_Vaishnav_complex_train_val.tsv": "c7b0964cd9132595961dbe28698e70c4",
        "url_Vaishnav_defined_drift.tsv": "https://zenodo.org/records/15195607/files/Vaishnav_defined_drift.tsv?download=1",
        "MD5_Vaishnav_defined_drift.tsv": "ff0cee1369303c0cc3469680027f432c",
        "url_Vaishnav_defined_native.tsv": "https://zenodo.org/records/15195607/files/Vaishnav_defined_native.tsv?download=1",
        "MD5_Vaishnav_defined_native.tsv": "95c6809a57a564d39e4a12be41682dd5",
        "url_Vaishnav_defined_paired.tsv": "https://zenodo.org/records/15195607/files/Vaishnav_defined_paired.tsv?download=1",
        "MD5_Vaishnav_defined_paired.tsv": "e156cf61cd413e31e73a60f41c8adc94",
        "url_Vaishnav_defined_train_val.tsv": "https://zenodo.org/records/15195607/files/Vaishnav_defined_train_val.tsv?download=1",
        "MD5_Vaishnav_defined_train_val.tsv": "1877098e135103dcc96e79c27bfa8122",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {},
        "task": "Regression",
        "split": {"train", "val", "test"},
        "Vaishnav_complex_split": {
            "train": 28214427,
            "val": 3134936,
            "test native": 3929,
            "test drift": 2983,
            "test paired": 2983,
        },
        "Vaishnav_defined_split": {
            "train": 18933667,
            "val": 2103740,
            "test native": 3978,
            "test drift": 2986,
            "test paired": 2986,
        },
    },
    "DeepStarr": {
        "python_class": "DeepStarr",
        "description": "The DeepSTARR data contains information on enhancer activity in D. melanogaster S2 cells. The dataset includes 282,892 sequences, each 249 nucleotides in length. The regression task is to predict two independent values of enhancer activity for two types of promoters.",
        "url_DeepStarr_all_chr.tsv": "https://zenodo.org/records/17453603/files/DeepStarr_all_chr.tsv?download=1",
        "MD5_DeepStarr_all_chr.tsv": "05addbb00e662cba478d13f1272e088d",
        "scalar_features": {},
        "vector_features": {},
        "target_columns": {"Dev_log2", "Hk_log2"},
        "task": "Regression",
        "split": {
            "train": "2L, 2LHet, 2RHet, 3L, 3LHet, 3R, 3RHet, 4, X, XHet, YHet",
            "val": "2R first half",
            "test": "2R second half",
        },
        "DeepStarr_split": {"train": 201139, "val": 40570, "test": 3978},
    },
}
